File: internal/engine/attacks.go
package engine

// IsSquareAttacked returns true if the given square is attacked by any piece of the specified color.
// This is an efficient check that works backwards from the target square to potential attackers.
func (b *Board) IsSquareAttacked(sq Square, byColor Color) bool {
	if !sq.IsValid() {
		return false
	}

	file := sq.File()
	rank := sq.Rank()

	// Check pawn attacks
	if b.isSquareAttackedByPawn(sq, file, rank, byColor) {
		return true
	}

	// Check knight attacks
	if b.isSquareAttackedByKnight(sq, file, rank, byColor) {
		return true
	}

	// Check king attacks
	if b.isSquareAttackedByKing(sq, file, rank, byColor) {
		return true
	}

	// Check diagonal attacks (bishop or queen)
	if b.isSquareAttackedDiagonally(file, rank, byColor) {
		return true
	}

	// Check orthogonal attacks (rook or queen)
	if b.isSquareAttackedOrthogonally(file, rank, byColor) {
		return true
	}

	return false
}

// isSquareAttackedByPawn checks if the square is attacked by a pawn of the given color.
// Pawns attack diagonally: white pawns attack upward (increasing rank),
// black pawns attack downward (decreasing rank).
func (b *Board) isSquareAttackedByPawn(sq Square, file, rank int, byColor Color) bool {
	// The pawn that could attack this square is one rank behind (from the attacker's perspective)
	// For white pawns attacking upward: they would be on rank-1
	// For black pawns attacking downward: they would be on rank+1
	var attackerRank int
	if byColor == White {
		attackerRank = rank - 1 // White pawn is below the target square
	} else {
		attackerRank = rank + 1 // Black pawn is above the target square
	}

	// Check bounds
	if attackerRank < 0 || attackerRank > 7 {
		return false
	}

	// Check both diagonal attack positions (file-1 and file+1)
	for _, attackerFile := range []int{file - 1, file + 1} {
		if attackerFile < 0 || attackerFile > 7 {
			continue
		}

		attackerSq := NewSquare(attackerFile, attackerRank)
		piece := b.Squares[attackerSq]
		if piece.Type() == Pawn && piece.Color() == byColor {
			return true
		}
	}

	return false
}

// isSquareAttackedByKnight checks if the square is attacked by a knight of the given color.
// Knights move in an L-shape: 2 squares in one direction, 1 square perpendicular.
func (b *Board) isSquareAttackedByKnight(sq Square, file, rank int, byColor Color) bool {
	// Knight move offsets: (file delta, rank delta)
	offsets := [][2]int{
		{+2, +1}, {+2, -1}, {-2, +1}, {-2, -1},
		{+1, +2}, {+1, -2}, {-1, +2}, {-1, -2},
	}

	for _, offset := range offsets {
		attackerFile := file + offset[0]
		attackerRank := rank + offset[1]

		// Check bounds
		if attackerFile < 0 || attackerFile > 7 || attackerRank < 0 || attackerRank > 7 {
			continue
		}

		attackerSq := NewSquare(attackerFile, attackerRank)
		piece := b.Squares[attackerSq]
		if piece.Type() == Knight && piece.Color() == byColor {
			return true
		}
	}

	return false
}

// isSquareAttackedByKing checks if the square is attacked by the king of the given color.
// Kings can attack any of the 8 adjacent squares.
func (b *Board) isSquareAttackedByKing(sq Square, file, rank int, byColor Color) bool {
	// King move offsets: all 8 adjacent squares
	offsets := [][2]int{
		{+1, +1}, {+1, -1}, {-1, +1}, {-1, -1}, // diagonal
		{+1, 0}, {-1, 0}, {0, +1}, {0, -1}, // orthogonal
	}

	for _, offset := range offsets {
		attackerFile := file + offset[0]
		attackerRank := rank + offset[1]

		// Check bounds
		if attackerFile < 0 || attackerFile > 7 || attackerRank < 0 || attackerRank > 7 {
			continue
		}

		attackerSq := NewSquare(attackerFile, attackerRank)
		piece := b.Squares[attackerSq]
		if piece.Type() == King && piece.Color() == byColor {
			return true
		}
	}

	return false
}

// isSquareAttackedDiagonally checks if the square is attacked by a bishop or queen diagonally.
// Slides along diagonals until hitting a piece or the board edge.
func (b *Board) isSquareAttackedDiagonally(file, rank int, byColor Color) bool {
	// Diagonal directions: (+1,+1), (+1,-1), (-1,+1), (-1,-1)
	directions := [][2]int{
		{+1, +1}, {+1, -1}, {-1, +1}, {-1, -1},
	}

	for _, dir := range directions {
		for dist := 1; dist <= 7; dist++ {
			attackerFile := file + dir[0]*dist
			attackerRank := rank + dir[1]*dist

			// Check bounds
			if attackerFile < 0 || attackerFile > 7 || attackerRank < 0 || attackerRank > 7 {
				break
			}

			attackerSq := NewSquare(attackerFile, attackerRank)
			piece := b.Squares[attackerSq]

			if piece.IsEmpty() {
				// Empty square, continue sliding
				continue
			}

			// Found a piece
			if piece.Color() == byColor {
				// It's the attacker's piece - check if it can attack diagonally
				if piece.Type() == Bishop || piece.Type() == Queen {
					return true
				}
			}
			// Either wrong color or a piece that can't attack diagonally - stop this direction
			break
		}
	}

	return false
}

// isSquareAttackedOrthogonally checks if the square is attacked by a rook or queen orthogonally.
// Slides along ranks and files until hitting a piece or the board edge.
func (b *Board) isSquareAttackedOrthogonally(file, rank int, byColor Color) bool {
	// Orthogonal directions: (+1,0), (-1,0), (0,+1), (0,-1)
	directions := [][2]int{
		{+1, 0}, {-1, 0}, {0, +1}, {0, -1},
	}

	for _, dir := range directions {
		for dist := 1; dist <= 7; dist++ {
			attackerFile := file + dir[0]*dist
			attackerRank := rank + dir[1]*dist

			// Check bounds
			if attackerFile < 0 || attackerFile > 7 || attackerRank < 0 || attackerRank > 7 {
				break
			}

			attackerSq := NewSquare(attackerFile, attackerRank)
			piece := b.Squares[attackerSq]

			if piece.IsEmpty() {
				// Empty square, continue sliding
				continue
			}

			// Found a piece
			if piece.Color() == byColor {
				// It's the attacker's piece - check if it can attack orthogonally
				if piece.Type() == Rook || piece.Type() == Queen {
					return true
				}
			}
			// Either wrong color or a piece that can't attack orthogonally - stop this direction
			break
		}
	}

	return false
}

File: rust/crates/engine/src/attacks.rs
//! Square-attack detection, working backwards from a target square to attackers.

use crate::board::Board;
use crate::types::{Color, PieceType, Square};

/// Knight move offsets: (file delta, rank delta).
const KNIGHT_OFFSETS: [(i32, i32); 8] = [
    (2, 1),
    (2, -1),
    (-2, 1),
    (-2, -1),
    (1, 2),
    (1, -2),
    (-1, 2),
    (-1, -2),
];

/// King move offsets: all 8 adjacent squares.
const KING_OFFSETS: [(i32, i32); 8] = [
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
];

/// Diagonal directions (bishop/queen).
const DIAGONAL_DIRS: [(i32, i32); 4] = [(1, 1), (1, -1), (-1, 1), (-1, -1)];

/// Orthogonal directions (rook/queen).
const ORTHOGONAL_DIRS: [(i32, i32); 4] = [(1, 0), (-1, 0), (0, 1), (0, -1)];

impl Board {
    /// Returns true if `sq` is attacked by any piece of `by_color`.
    pub fn is_square_attacked(&self, sq: Square, by_color: Color) -> bool {
        if !sq.is_valid() {
            return false;
        }

        let file = sq.file();
        let rank = sq.rank();

        self.is_square_attacked_by_pawn(file, rank, by_color)
            || self.is_square_attacked_by_knight(file, rank, by_color)
            || self.is_square_attacked_by_king(file, rank, by_color)
            || self.is_square_attacked_diagonally(file, rank, by_color)
            || self.is_square_attacked_orthogonally(file, rank, by_color)
    }

    fn is_square_attacked_by_pawn(&self, file: i32, rank: i32, by_color: Color) -> bool {
        // A pawn attacking this square sits one rank behind (from its perspective).
        let attacker_rank = if by_color == Color::White {
            rank - 1
        } else {
            rank + 1
        };

        if !(0..=7).contains(&attacker_rank) {
            return false;
        }

        for attacker_file in [file - 1, file + 1] {
            if !(0..=7).contains(&attacker_file) {
                continue;
            }
            let attacker_sq = Square::new(attacker_file, attacker_rank);
            let piece = self.squares[attacker_sq.index()];
            if piece.piece_type() == PieceType::Pawn && piece.color() == by_color {
                return true;
            }
        }
        false
    }

    fn is_square_attacked_by_knight(&self, file: i32, rank: i32, by_color: Color) -> bool {
        for (df, dr) in KNIGHT_OFFSETS {
            let af = file + df;
            let ar = rank + dr;
            if !(0..=7).contains(&af) || !(0..=7).contains(&ar) {
                continue;
            }
            let attacker_sq = Square::new(af, ar);
            let piece = self.squares[attacker_sq.index()];
            if piece.piece_type() == PieceType::Knight && piece.color() == by_color {
                return true;
            }
        }
        false
    }

    fn is_square_attacked_by_king(&self, file: i32, rank: i32, by_color: Color) -> bool {
        for (df, dr) in KING_OFFSETS {
            let af = file + df;
            let ar = rank + dr;
            if !(0..=7).contains(&af) || !(0..=7).contains(&ar) {
                continue;
            }
            let attacker_sq = Square::new(af, ar);
            let piece = self.squares[attacker_sq.index()];
            if piece.piece_type() == PieceType::King && piece.color() == by_color {
                return true;
            }
        }
        false
    }

    fn is_square_attacked_diagonally(&self, file: i32, rank: i32, by_color: Color) -> bool {
        for (df, dr) in DIAGONAL_DIRS {
            for dist in 1..=7 {
                let af = file + df * dist;
                let ar = rank + dr * dist;
                if !(0..=7).contains(&af) || !(0..=7).contains(&ar) {
                    break;
                }
                let attacker_sq = Square::new(af, ar);
                let piece = self.squares[attacker_sq.index()];
                if piece.is_empty() {
                    continue;
                }
                if piece.color() == by_color
                    && (piece.piece_type() == PieceType::Bishop
                        || piece.piece_type() == PieceType::Queen)
                {
                    return true;
                }
                break;
            }
        }
        false
    }

    fn is_square_attacked_orthogonally(&self, file: i32, rank: i32, by_color: Color) -> bool {
        for (df, dr) in ORTHOGONAL_DIRS {
            for dist in 1..=7 {
                let af = file + df * dist;
                let ar = rank + dr * dist;
                if !(0..=7).contains(&af) || !(0..=7).contains(&ar) {
                    break;
                }
                let attacker_sq = Square::new(af, ar);
                let piece = self.squares[attacker_sq.index()];
                if piece.is_empty() {
                    continue;
                }
                if piece.color() == by_color
                    && (piece.piece_type() == PieceType::Rook
                        || piece.piece_type() == PieceType::Queen)
                {
                    return true;
                }
                break;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Piece;

    fn empty() -> Board {
        Board::default()
    }

    fn place(b: &mut Board, file: i32, rank: i32, color: Color, pt: PieceType) {
        b.squares[Square::new(file, rank).index()] = Piece::new(color, pt);
    }

    #[test]
    fn empty_board_no_attacks() {
        let board = empty();
        let e4 = Square::new(4, 3);
        assert!(!board.is_square_attacked(e4, Color::White));
        assert!(!board.is_square_attacked(e4, Color::Black));
    }

    #[test]
    fn invalid_square_returns_false() {
        let board = Board::new();
        assert!(!board.is_square_attacked(crate::types::NO_SQUARE, Color::White));
        assert!(!board.is_square_attacked(Square(-5), Color::Black));
        assert!(!board.is_square_attacked(Square(64), Color::White));
    }

    #[test]
    fn pawn_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Pawn); // e4
        assert!(board.is_square_attacked(Square::new(3, 4), Color::White)); // d5
        assert!(board.is_square_attacked(Square::new(5, 4), Color::White)); // f5
        assert!(!board.is_square_attacked(Square::new(4, 4), Color::White)); // e5 forward
        assert!(!board.is_square_attacked(Square::new(3, 2), Color::White)); // d3
        assert!(!board.is_square_attacked(Square::new(5, 2), Color::White)); // f3

        let mut board = empty();
        place(&mut board, 4, 4, Color::Black, PieceType::Pawn); // e5
        assert!(board.is_square_attacked(Square::new(3, 3), Color::Black)); // d4
        assert!(board.is_square_attacked(Square::new(5, 3), Color::Black)); // f4
        assert!(!board.is_square_attacked(Square::new(4, 3), Color::Black)); // e4 forward

        // a-file pawn only attacks right
        let mut board = empty();
        place(&mut board, 0, 1, Color::White, PieceType::Pawn); // a2
        assert!(board.is_square_attacked(Square::new(1, 2), Color::White)); // b3

        // wrong color
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Pawn);
        assert!(board.is_square_attacked(Square::new(3, 4), Color::White));
        assert!(!board.is_square_attacked(Square::new(3, 4), Color::Black));
    }

    #[test]
    fn knight_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Knight); // e4
        for (f, r) in [
            (5, 5),
            (6, 4),
            (6, 2),
            (5, 1),
            (3, 1),
            (2, 2),
            (2, 4),
            (3, 5),
        ] {
            assert!(board.is_square_attacked(Square::new(f, r), Color::White));
        }
        assert!(!board.is_square_attacked(Square::new(4, 2), Color::White));
        assert!(!board.is_square_attacked(Square::new(4, 4), Color::White));

        let mut board = empty();
        place(&mut board, 0, 0, Color::White, PieceType::Knight); // a1
        assert!(board.is_square_attacked(Square::new(1, 2), Color::White)); // b3
        assert!(board.is_square_attacked(Square::new(2, 1), Color::White)); // c2
        assert!(!board.is_square_attacked(Square::new(0, 2), Color::White)); // a3
    }

    #[test]
    fn bishop_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Bishop); // e4
        for (f, r) in [
            (5, 4),
            (7, 6),
            (5, 2),
            (7, 0),
            (3, 2),
            (1, 0),
            (3, 4),
            (0, 7),
        ] {
            assert!(board.is_square_attacked(Square::new(f, r), Color::White));
        }
        assert!(!board.is_square_attacked(Square::new(4, 4), Color::White));
        assert!(!board.is_square_attacked(Square::new(3, 3), Color::White));

        // blocked by own piece
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Bishop);
        place(&mut board, 5, 4, Color::White, PieceType::Knight);
        assert!(board.is_square_attacked(Square::new(5, 4), Color::White));
        assert!(!board.is_square_attacked(Square::new(6, 5), Color::White));
        assert!(!board.is_square_attacked(Square::new(7, 6), Color::White));
        assert!(board.is_square_attacked(Square::new(3, 4), Color::White));

        // corner bishop
        let mut board = empty();
        place(&mut board, 0, 0, Color::White, PieceType::Bishop);
        assert!(board.is_square_attacked(Square::new(7, 7), Color::White));
        assert!(board.is_square_attacked(Square::new(3, 3), Color::White));
        assert!(!board.is_square_attacked(Square::new(1, 0), Color::White));
    }

    #[test]
    fn rook_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Rook); // e4
        for (f, r) in [
            (4, 4),
            (4, 7),
            (4, 2),
            (4, 0),
            (5, 3),
            (7, 3),
            (3, 3),
            (0, 3),
        ] {
            assert!(board.is_square_attacked(Square::new(f, r), Color::White));
        }
        assert!(!board.is_square_attacked(Square::new(5, 4), Color::White));
        assert!(!board.is_square_attacked(Square::new(3, 2), Color::White));

        // blocked
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Rook);
        place(&mut board, 4, 5, Color::White, PieceType::Pawn); // e6
        assert!(board.is_square_attacked(Square::new(4, 4), Color::White)); // e5
        assert!(board.is_square_attacked(Square::new(4, 5), Color::White)); // e6
        assert!(!board.is_square_attacked(Square::new(4, 6), Color::White)); // e7
        assert!(!board.is_square_attacked(Square::new(4, 7), Color::White)); // e8

        // blocked by enemy, still attacks the enemy square
        let mut board = empty();
        place(&mut board, 0, 0, Color::White, PieceType::Rook); // a1
        place(&mut board, 0, 3, Color::Black, PieceType::Pawn); // a4
        assert!(board.is_square_attacked(Square::new(0, 3), Color::White));
        assert!(!board.is_square_attacked(Square::new(0, 7), Color::White));
    }

    #[test]
    fn queen_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::Queen); // e4
        assert!(board.is_square_attacked(Square::new(7, 6), Color::White)); // h7
        assert!(board.is_square_attacked(Square::new(0, 7), Color::White)); // a8
        assert!(board.is_square_attacked(Square::new(4, 7), Color::White)); // e8
        assert!(board.is_square_attacked(Square::new(0, 3), Color::White)); // a4
    }

    #[test]
    fn king_attacks() {
        let mut board = empty();
        place(&mut board, 4, 3, Color::White, PieceType::King); // e4
        for (f, r) in [
            (3, 2),
            (4, 2),
            (5, 2),
            (3, 3),
            (5, 3),
            (3, 4),
            (4, 4),
            (5, 4),
        ] {
            assert!(board.is_square_attacked(Square::new(f, r), Color::White));
        }
        assert!(!board.is_square_attacked(Square::new(4, 5), Color::White));
        assert!(!board.is_square_attacked(Square::new(6, 5), Color::White));

        let mut board = empty();
        place(&mut board, 0, 0, Color::White, PieceType::King);
        assert!(board.is_square_attacked(Square::new(0, 1), Color::White));
        assert!(board.is_square_attacked(Square::new(1, 0), Color::White));
        assert!(board.is_square_attacked(Square::new(1, 1), Color::White));
        assert!(!board.is_square_attacked(Square::new(0, 2), Color::White));
    }

    #[test]
    fn starting_position_control() {
        let board = Board::new();
        assert!(board.is_square_attacked(Square::new(4, 2), Color::White)); // e3
        assert!(board.is_square_attacked(Square::new(4, 5), Color::Black)); // e6
        assert!(board.is_square_attacked(Square::new(0, 2), Color::White)); // a3 knight
        assert!(board.is_square_attacked(Square::new(2, 2), Color::White)); // c3 knight
        assert!(!board.is_square_attacked(Square::new(4, 3), Color::White)); // e4
        assert!(!board.is_square_attacked(Square::new(4, 3), Color::Black));
    }
}
